azure/functions/_http_asgi.py

```diff
@@ -0,0 +1,134 @@
+# Licensed under the MIT License.
+
+import asyncio
+from typing import Callable, Dict, List, Tuple, Optional, Any, Union
+import logging
+from wsgiref.headers import Headers
+
+from ._abc import Context
+from ._http import HttpRequest, HttpResponse
+from ._http_wsgi import WsgiRequest
+
+
+class AsgiRequest(WsgiRequest):
+    _environ_cache: Optional[Dict[str, Any]] = None
+
+    def __init__(self, func_req: HttpRequest,
+                 func_ctx: Optional[Context] = None):
+        self.asgi_version = "2.1"
+        self.asgi_spec_version = "2.1"
+        self._headers = func_req.headers
+        super().__init__(func_req, func_ctx)
+
+    def _get_encoded_http_headers(self) -> List[Tuple[bytes, bytes]]:
+        return [(k.encode("utf8"), v.encode("utf8"))
+                for k, v in self._headers.items()]
+
+    def _get_server_address(self):
+        if self.server_name is not None:
+            return (self.server_name, int(self.server_port))
+        return None
+
+    def to_asgi_http_scope(self):
+        return {
+            "type": "http",
+            "asgi.version": self.asgi_version,
+            "asgi.spec_version": self.asgi_spec_version,
+            "http_version": "1.1",
+            "method": self.request_method,
+            "scheme": "https",
+            "path": self.path_info,
+            "raw_path": self.path_info.encode("utf-8"),
+            "query_string": self.query_string.encode("utf-8"),
+            "root_path": self.script_name,
+            "headers": self._get_encoded_http_headers(),
+            "server": self._get_server_address(),
+            "client": None
+        }
+        # Notes, missing client name, port
+
+
+class AsgiResponse:
+    def __init__(self):
+        self._status_code = 0
+        self._headers: Union[Headers, Dict] = {}
+        self._buffer: List[bytes] = []
+        self._request_body = b""
+
+    @classmethod
+    async def from_app(cls, app, scope: Dict[str, Any],
+                       body: bytes) -> "AsgiResponse":
+        res = cls()
+        res._request_body = body
+        await app(scope, res._receive, res._send)
+        return res
+
+    def to_func_response(self) -> HttpResponse:
+        lowercased_headers = {k.lower(): v for k, v in self._headers.items()}
+        return HttpResponse(
+            body=b"".join(self._buffer),
+            status_code=self._status_code,
+            headers=self._headers,
+            mimetype=lowercased_headers.get("content-type"),
+            charset=lowercased_headers.get("content-encoding"),
+        )
+
+    def _handle_http_response_start(self, message: Dict[str, Any]):
+        self._headers = Headers(
+            [(k.decode(), v.decode())
+             for k, v in message["headers"]])
+        self._status_code = message["status"]
+
+    def _handle_http_response_body(self, message: Dict[str, Any]):
+        self._buffer.append(message["body"])
+        # TODO : Handle more_body flag
```

Comment:
I don't think we support this feature in the host yet, so please to update the TODO comment and link to this issue https://github.com/Azure/azure-functions-host/issues/4926. Basically, we don't support chunked HTTP body, because it is based on the data stream protocol.